# 01 - Read and Join Data

## Import Libraries and Connect to PostgreSQL

In [2]:
import numpy as np
import pandas as pd
from sqlalchemy import create_engine

In [1]:
from pathlib import Path

PROJECT_ROOT = Path.cwd().parent

**Connect with dataase in PostgreSQL**

In [2]:
engine = create_engine("postgresql+psycopg2://postgres:1234@localhost:5432/olist")

In [5]:
# Test the connection
query = "SELECT COUNT(*) AS total_orders FROM orders;"
pd.read_sql(query, engine)

,total_orders
0,99441


### Read Tables from PostgreSQL

In [6]:
orders = pd.read_sql("SELECT * FROM orders", engine)

customers = pd.read_sql("SELECT * FROM customers", engine)

order_items = pd.read_sql("SELECT * FROM order_items", engine)

order_payments = pd.read_sql("SELECT * FROM order_payments", engine)

order_reviews = pd.read_sql("SELECT * FROM order_reviews", engine)

products = pd.read_sql("SELECT * FROM products", engine)

sellers = pd.read_sql("SELECT * FROM sellers", engine)

geolocation = pd.read_sql("SELECT * FROM geolocation", engine)

category_translation = pd.read_sql("SELECT * FROM category_translation", engine)

**Note:** 

`category_translation` was inspected but not joined because product category names were not selected for the final order-level prediction table.

### Inspect the Tables

In [7]:
print("orders:", orders.shape)
print("customers:", customers.shape)
print("order_items:", order_items.shape)
print("order_payments:", order_payments.shape)
print("order_reviews:", order_reviews.shape)
print("products:", products.shape)
print("sellers:", sellers.shape)
print("geolocation:", geolocation.shape)

orders: (99441, 8)
customers: (99441, 5)
order_items: (112650, 7)
order_payments: (103886, 5)
order_reviews: (99224, 7)
products: (32951, 9)
sellers: (3095, 4)
geolocation: (1000163, 5)


In [8]:
table_checks = {
    "orders": (orders, ["order_id"]),
    "customers": (customers, ["customer_id"]),
    "order_items": (order_items, ["order_id", "order_item_id"]),
    "order_payments": (order_payments, ["order_id", "payment_sequential"]),
    "order_reviews": (order_reviews, ["review_id"]),
    "products": (products, ["product_id"]),
    "sellers": (sellers, ["seller_id"])
}

for name, (df, key_cols) in table_checks.items():
    print(f"\n{name}")
    print("Rows:", len(df))
    print("Duplicate key rows:", df.duplicated(subset=key_cols).sum())


orders
Rows: 99441
Duplicate key rows: 0

customers
Rows: 99441
Duplicate key rows: 0

order_items
Rows: 112650
Duplicate key rows: 0

order_payments
Rows: 103886
Duplicate key rows: 0

order_reviews
Rows: 99224
Duplicate key rows: 814

products
Rows: 32951
Duplicate key rows: 0

sellers
Rows: 3095
Duplicate key rows: 0


## Aggregate before joining

* **order_items**

In [9]:
print(order_items.head())

                           order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4

In [10]:
items_agg = order_items.groupby("order_id").agg(
    total_price=("price", "sum"),
    total_freight=("freight_value", "sum"),
    item_count=("order_item_id", "count")
).reset_index()

items_agg.head()

,order_id,total_price,total_freight,item_count
0,00010242fe8c5a6d1ba2dd792cb16214,58.90,13.29,1
1,00018f77f2f0320c557190d7a144bdd3,239.90,19.93,1
2,000229ec398224ef6ca0657da4fc703e,199.00,17.87,1
3,00024acbcdf0a6daa1e931b038114c75,12.99,12.79,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,199.90,18.14,1


* **order_payments** 

In [11]:
print(order_payments.head())

                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  


In [12]:
payments_agg = order_payments.groupby("order_id").agg(
    total_payment=("payment_value", "sum"),
    payment_installments=("payment_installments", "max"),
    payment_count=("payment_sequential", "count")
).reset_index()

payments_agg.head()

,order_id,total_payment,payment_installments,payment_count
0,00010242fe8c5a6d1ba2dd792cb16214,72.19,2,1
1,00018f77f2f0320c557190d7a144bdd3,259.83,3,1
2,000229ec398224ef6ca0657da4fc703e,216.87,5,1
3,00024acbcdf0a6daa1e931b038114c75,25.78,2,1
4,00042b26cf59d7ce69dfabb4e55b4fd9,218.04,3,1


* **Reviews** 

In [13]:
print(order_reviews.head())

                          review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN           2018-01-18   
1                                                NaN           2018-03-10   
2                                                NaN           2018-02-17   
3           

In [14]:
reviews_agg = order_reviews.groupby("order_id").agg(
    review_score=("review_score", "mean")
).reset_index()

reviews_agg.head()

,order_id,review_score
0,00010242fe8c5a6d1ba2dd792cb16214,5.0
1,00018f77f2f0320c557190d7a144bdd3,4.0
2,000229ec398224ef6ca0657da4fc703e,5.0
3,00024acbcdf0a6daa1e931b038114c75,4.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,5.0


## Merge Tables (Joining)

* orders + customers

In [15]:
data = orders.merge(customers, on="customer_id", how='left')
data.shape

(99441, 12)

* orders + items_agg (previously order_items)

In [16]:
data = data.merge(items_agg, on= "order_id", how='left')
data.shape

(99441, 15)

* orders + payments_agg (previously order_payments)

In [17]:
data = data.merge(payments_agg, on= "order_id", how='left')
data.shape

(99441, 18)

* orders + reviews_agg (previously order_reviews)

In [18]:
data = data.merge(reviews_agg, on= "order_id", how='left')
data.shape

(99441, 19)

* **seller**

In [19]:
# Add seller information to order items
items_with_sellers = order_items.merge(
    sellers[["seller_id", "seller_zip_code_prefix", "seller_city", "seller_state"]],
    
    on="seller_id", how="left")

items_with_sellers.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,seller_zip_code_prefix,seller_city,seller_state
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,27277,volta redonda,SP
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,3471,sao paulo,SP
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,37564,borda da mata,MG
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,14403,franca,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,87900,loanda,PR


In [20]:
seller_agg = items_with_sellers.groupby("order_id").agg(
    seller_count=("seller_id", "nunique"),
    seller_state_count=("seller_state", "nunique"),
    seller_states=("seller_state", lambda x: ",".join(sorted(x.dropna().unique()))),
    seller_zip_code_prefix=("seller_zip_code_prefix", "min")
).reset_index()

seller_agg.head()

,order_id,seller_count,seller_state_count,seller_states,seller_zip_code_prefix
0,00010242fe8c5a6d1ba2dd792cb16214,1,1,SP,27277
1,00018f77f2f0320c557190d7a144bdd3,1,1,SP,3471
2,000229ec398224ef6ca0657da4fc703e,1,1,MG,37564
3,00024acbcdf0a6daa1e931b038114c75,1,1,SP,14403
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,1,PR,87900


In [21]:
print("Seller aggregation shape:", seller_agg.shape)
print("Unique orders:", seller_agg["order_id"].nunique())
print("Duplicated orders:", seller_agg["order_id"].duplicated().sum())

Seller aggregation shape: (98666, 5)
Unique orders: 98666
Duplicated orders: 0


In [22]:
# Merge Seller Information
data = data.merge(seller_agg, on="order_id", how="left")

print("Dataset shape after seller merge:", data.shape)
print("Unique orders:", data["order_id"].nunique())
print("Duplicated orders:", data["order_id"].duplicated().sum())

Dataset shape after seller merge: (99441, 23)
Unique orders: 99441
Duplicated orders: 0


* **Geolocation**

In [23]:
geo_agg = geolocation.groupby("geolocation_zip_code_prefix").agg(
    lat=("geolocation_lat", "mean"),
    lng=("geolocation_lng", "mean")
).reset_index()

print("Unique ZIP locations:", geo_agg.shape)
print("Duplicated ZIP prefixes:", geo_agg["geolocation_zip_code_prefix"].duplicated().sum())

geo_agg.head()

Unique ZIP locations: (19015, 3)
Duplicated ZIP prefixes: 0


,geolocation_zip_code_prefix,lat,lng
0,1001,-23.550190,-46.634024
1,1002,-23.548146,-46.634979
2,1003,-23.548994,-46.635731
3,1004,-23.549799,-46.634757
4,1005,-23.549456,-46.636733


**Customer Coordinates**

In [24]:
customer_geo = geo_agg.rename(columns={
    "geolocation_zip_code_prefix": "customer_zip_code_prefix",
    "lat": "customer_lat",
    "lng": "customer_lng"
})

data = data.merge(
    customer_geo,
    on="customer_zip_code_prefix",
    how="left"
)

print("Shape after customer coordinates:", data.shape)
print("Duplicated orders:", data["order_id"].duplicated().sum())

Shape after customer coordinates: (99441, 25)
Duplicated orders: 0


**Seller Coordinates**

In [25]:
seller_geo_coords = geo_agg.rename(columns={
    "geolocation_zip_code_prefix": "seller_zip_code_prefix",
    "lat": "seller_lat",
    "lng": "seller_lng"
})

data = data.merge(
    seller_geo_coords,
    on="seller_zip_code_prefix",
    how="left"
)

print("Shape after seller coordinates:", data.shape)
print("Duplicated orders:", data["order_id"].duplicated().sum())

data[
    ["seller_zip_code_prefix", "seller_lat", "seller_lng"]
].head()

Shape after seller coordinates: (99441, 27)
Duplicated orders: 0


,seller_zip_code_prefix,seller_lat,seller_lng
0,9350.0,-23.680729,-46.444238
1,31570.0,-19.807681,-43.980427
2,14840.0,-21.363502,-48.229601
3,31842.0,-19.837682,-43.924053
4,8752.0,-23.543395,-46.262086


**distance_km**

In [26]:
# Initialize distance as missing
data["distance_km"] = np.nan

# Distance is calculated only for single-seller orders
valid_distance = (
    (data["seller_count"] == 1)
    & data["customer_lat"].notna()
    & data["customer_lng"].notna()
    & data["seller_lat"].notna()
    & data["seller_lng"].notna())

# Convert coordinates to radians
lat1 = np.radians(data.loc[valid_distance, "customer_lat"])
lon1 = np.radians(data.loc[valid_distance, "customer_lng"])
lat2 = np.radians(data.loc[valid_distance, "seller_lat"])
lon2 = np.radians(data.loc[valid_distance, "seller_lng"])

# Haversine formula
dlat = lat2 - lat1
dlon = lon2 - lon1

a = (np.sin(dlat / 2) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2) ** 2)

c = 2 * np.arcsin(np.sqrt(a))

data.loc[valid_distance, "distance_km"] = 6371 * c

print("Orders with valid distance:", data["distance_km"].notna().sum())
print("Orders without distance:", data["distance_km"].isna().sum())

data["distance_km"].describe()

Orders with valid distance: 96905
Orders without distance: 2536


count    96905.000000
mean       602.297902
std        596.301543
min          0.000000
25%        184.858896
50%        433.882900
75%        800.033462
max       8677.911622
Name: distance_km, dtype: float64

In [27]:
print("Dataset shape:", data.shape)
print("Unique orders:", data["order_id"].nunique())
print("Duplicated orders:", data["order_id"].duplicated().sum())

data.head()

Dataset shape: (99441, 28)
Unique orders: 99441
Duplicated orders: 0


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,review_score,seller_count,seller_state_count,seller_states,seller_zip_code_prefix,customer_lat,customer_lng,seller_lat,seller_lng,distance_km
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,4.0,1.0,1.0,SP,9350.0,-23.576983,-46.587161,-23.680729,-46.444238,18.576110
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,4.0,1.0,1.0,SP,31570.0,-12.177924,-44.660711,-19.807681,-43.980427,851.495069
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,5.0,1.0,1.0,SP,14840.0,-16.745150,-48.514783,-21.363502,-48.229601,514.410666
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,...,5.0,1.0,1.0,MG,31842.0,-5.774190,-35.271143,-19.837682,-43.924053,1822.226336
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,5.0,1.0,1.0,SP,8752.0,-23.676370,-46.514627,-23.543395,-46.262086,29.676625


## Missing Values Check

**After merging all tables, we checked the dataset for missing values.**

- The missing values are not necessarily errors. Some of them are expected because certain information is only available depending on the order status or whether a related record exists in another table. Let's check that:

In [28]:
missing = data.isnull().sum()
missing[missing > 0].sort_values(ascending=False)

order_delivered_customer_date    2965
distance_km                      2536
order_delivered_carrier_date     1783
seller_lng                        992
seller_lat                        992
seller_states                     775
total_price                       775
seller_state_count                775
total_freight                     775
item_count                        775
seller_count                      775
seller_zip_code_prefix            775
review_score                      768
customer_lng                      278
customer_lat                      278
order_approved_at                 160
payment_installments                1
total_payment                       1
payment_count                       1
dtype: int64

In [29]:
# check order_delivered_customer_date    
data.groupby("order_status").size().sort_values(ascending=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
dtype: int64

In [30]:
data[data["order_delivered_customer_date"].isnull()] \
    .groupby("order_status").size() \
    .sort_values(ascending=False)

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
dtype: int64

Most missing delivery dates belong to orders that were not delivered.
Next, we check whether any delivered orders are also missing the delivery date.

In [31]:
delivered_missing = data[
    (data["order_status"] == "delivered") &
    (data["order_delivered_customer_date"].isnull())
]

print("Delivered orders with missing delivery date:", len(delivered_missing))

delivered_missing[
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

Delivered orders with missing delivery date: 8


,order_id,order_status,order_purchase_timestamp,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,delivered,2017-11-28 17:44:07,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,delivered,2018-06-20 06:58:43,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,delivered,2018-07-01 17:05:11,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,delivered,2018-07-01 22:05:55,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,delivered,2018-07-01 21:14:02,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,delivered,2017-05-25 23:22:43,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,delivered,2018-06-08 12:09:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,delivered,2018-06-27 16:09:12,2018-07-03 19:26:00,NaT,2018-07-19


There are 8 delivered orders with a missing delivery date.
These records are kept unchanged because the actual delivery date cannot be inferred reliably.

In [32]:
data[data["item_count"].isnull()] \
    .groupby("order_status").size() \
    .sort_values(ascending=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
dtype: int64

In [33]:
data[data["review_score"].isnull()] \
    .groupby("order_status").size() \
    .sort_values(ascending=False)

order_status
delivered      646
shipped         75
canceled        20
unavailable     14
processing       6
invoiced         5
created          2
dtype: int64

In [34]:
data[data["total_payment"].isnull()][
    ["order_id", "order_status", "total_price", "total_payment"]
]

,order_id,order_status,total_price,total_payment
30710,bfbd0f9bdef84302105ad712db648a6c,delivered,134.97,NaN


In [35]:
missing_payment_order = data.loc[
    data['total_payment'].isnull(), "order_id"].iloc[0]

order_payments[order_payments['order_id'] == missing_payment_order]

,order_id,payment_sequential,payment_type,payment_installments,payment_value


Final Check

In [36]:
print("Final dataset shape:", data.shape)
print("Unique orders:", data["order_id"].nunique())
print("Duplicated orders:", data["order_id"].duplicated().sum())

print("\nMissing values:")
print(data.isnull().sum().loc[lambda x: x > 0].sort_values(ascending=False))

Final dataset shape: (99441, 28)
Unique orders: 99441
Duplicated orders: 0

Missing values:
order_delivered_customer_date    2965
distance_km                      2536
order_delivered_carrier_date     1783
seller_lng                        992
seller_lat                        992
seller_states                     775
total_price                       775
seller_state_count                775
total_freight                     775
item_count                        775
seller_count                      775
seller_zip_code_prefix            775
review_score                      768
customer_lng                      278
customer_lat                      278
order_approved_at                 160
payment_installments                1
total_payment                       1
payment_count                       1
dtype: int64


#### **Investigation of Missing Values**

We investigated the missing values instead of filling or dropping them immediately.

- `order_delivered_customer_date`: Missing mainly for orders that were not successfully delivered.
- `order_delivered_carrier_date`: Missing when the order was not shipped or had not reached the carrier stage.
- `review_score`: Some orders do not have a corresponding review.
- `total_price`, `total_freight`, and `item_count`: Missing when no matching order items were found.
- Seller-related columns: Missing when no matching order items or seller information were available.
- Customer and seller coordinates: Missing when the ZIP code was not available in the geolocation table.
- `distance_km`: Missing when coordinates were unavailable or when an order had multiple sellers.
- Payment-related columns: One order has no matching record in the payments table.

Most missing values are therefore related to order status or unavailable records in the related tables rather than random missing data.

---

### **Missing Payment Investigation**

- Only one order has missing payment information.

- We checked the original `order_payments` table using its `order_id` and found no matching payment record.

- Since a left join was used, the order remains in the merged dataset even though no corresponding payment record exists.

- The missing payment values were kept as `NaN` rather than replaced with `0`, because missing payment information does not necessarily indicate a zero payment amount.

---

In [ ]:
data.to_csv(PROJECT_ROOT / "data" / "merged_orders.csv",index=False)

print("Saved merged_orders.csv")
print("Final shape:", data.shape)

Saved merged_orders.csv
Final shape: (99441, 28)


### Summary

In this notebook, the required tables were loaded from PostgreSQL and aggregated to maintain one row per order.

Customer, item, payment, review, seller, and geolocation information were merged into the main dataset. Customer-seller distance was calculated for single-seller orders using the Haversine formula.

The final dataset was checked for duplicate orders and missing values. Missing values were investigated and kept when they represented unavailable information rather than being replaced arbitrarily.

The final merged dataset contains 99,441 unique orders and is saved as `merged_orders.csv` for the next steps of the project.